accepted/2025/audit-sources-visual-studio-options.md

```diff
@@ -0,0 +1,140 @@
+- GitHub Issue: <https://github.com/nuget/home/issues/14583>
+
+## **Summary**
+
+Visual Studio currently supports NuGet Audit to help developers identify vulnerable packages. As adoption grows, customers need a convenient way to manage **Audit Sources** directly from the Visual Studio Options UI, rather than editing `nuget.config` manually. This proposal introduces first-class support for adding, removing, and updating Audit Sources in Visual Studio.
+
+## **Motivation**
+
+Pain-points today:
+
+- Customers must manually edit `NuGet.Config` to configure `<auditSources>`.
+  There can be multiple files containing this configuration.
+- Lack of discoverability for audit-related settings in Visual Studio.
+
+## **Explanation**
+
+### **Functional Explanation**
+
+Add an **"Audit Sources"** table to the "Package Sources" page in NuGet's Visual Studio Options.
+
+- **Discoverability**: Enable Quick Search (Ctrl+Q) so that searching for "Audit Source" navigates to the Package Sources page in Unified Settings.
+- Up to **Three Tables** can be shown in this order:
+  - Package Sources (always shown)
+  - Audit Sources (shown when explicitly configured),
+  - Machine-wide Package Sources (shown when explicitly configured)
+- New dropdown: **Choose how NuGet Audit retrieves vulnerability data**
+  - Introduce a dropdown control to switch from **"Read vulnerabilities from my package sources"** to **"Configure sources to read vulnerabilities"**
+  - When an audit source is configured, the Audit Sources table appears.
+  - To configure the first audit source, select **"Configure sources to read vulnerabilities"**; this reveals the Audit Sources table.
+
+#### Read vulnerabilities from my package sources
+
+- Default option - if no audit sources exist, this will be the selection.
+- Audit Sources table will be **hidden**.
+
+![Existing package sources Visual Studio Options page shown with a new "Choose how NuGet Audit retrieves vulnerability data" dropdown](../../meta/resources/AuditSources-VS/audit-choose-package-source-page.png)
+
+#### Configure sources to read vulnerabilities
+
+- Audit Sources table will be **shown**.
+- **Pre-selected**  when one or more audit sources are already configured.
+  - The dropdown will be disabled as well since the presence of audit sources takes away the behavior of reading vulnerability data from package sources.
+- **User-selectable** only when no audit sources exist, enabling  a customer to explicitly configure their first `<auditSource>` using the `Add` button.
+
+![New dropdown expanded with both options shown"](../../meta/resources/AuditSources-VS/choose-nuget-audit-dropdown.png)
+
+![Table of Audit Sources shown below Package Sources because Combobox option "Configure sources to read vulnerabilities" is selected](../../meta/resources/AuditSources-VS/audit-sources-visible-table.png)
+
+- Switching the dropdown back to "**Read vulnerabilities from my package sources**" would not be supported in this iteration.
```

Comment:
Why not? Seems straightforward to implement.

Reply:
NuGet Restore will use audit sources if they exist. 
What isn't straightforward is making the uncheck automatically remove all of those audit sources. 
I don't expect more than an average of 1 audit source, maybe 2-3 audit sources in the more extreme cases. It's also probably less common to go from audit sources back to package source vulnerability resources.

Therefore, customers can press Remove on all their audit sources and end up back in the original state. Additionally, the Unified Settings team is exploring building in a Remove All button on array settings.

Reply:
In that case, could we simplify the UX by always showing a list of Audit Sources (even if empty)? Having a dropdown that only works for opting in but not for opting out seems a bit broken.

Reply:
We discussed this option first, and we believe it's confusing to have an empty table taking up space that isn't used by Audit in most cases. We agreed on hiding it entirely behind a user's choice to explicitly configure Audit. This makes it a tangible act that you're trying to manage your audit sources yourself.

It'll be a Checkbox instead of a Dropdown, as mentioned - but it doesn't opt you in to anything. It just enables you to choose audit sources. We do not have any setting today that opts you in to using audit sources; we use them if they are present.

Reply:
I like the part of the spec on the next line that covers how to revert back to using only package sources as the audit sources:
> If customers want the ability to switch back from their audit sources to only package sources for Vulnerability data, a future iteration could support this and automatically clear `<auditSources>` after showing a warning messagebox that can be cancelled.

I think this would be a great addition for the first iteration. Allowing users to clear their audit sources directly in the UI creates a complete experience. Is there a way to programmatically iterate over the list of audit sources and remove each one from the nuget.config file(s) when the user confirms toggles the checkbox and confirms in the dialog?